File: android/app/src/main/java/dev/pam/nativeapp/modules/SensorsModule.kt
package dev.pam.nativeapp.modules

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue
import java.util.concurrent.atomic.AtomicInteger

internal class SensorsModule(context: Context) : NativeModule, AutoCloseable {
    private val manager = context.getSystemService(SensorManager::class.java)
    private val nextId = AtomicInteger(1)
    private val watches = HashMap<Int, Watch>()

    override fun invoke(method: String, payload: ByteArray, completion: ModuleCompletion) {
        runCatching {
            when (method) {
                "watch" -> start(payload, completion)
                "next" -> watch(payload).channel.next(completion)
                "stop" -> {
                    stop(id(payload))
                    completion.complete(ModuleResultStatus.SUCCESS, ByteArray(0))
                }
                else -> error("Unknown sensors method $method")
            }
        }.onFailure {
            completion.complete(ModuleResultStatus.FAILURE, (it.message ?: "Sensor observation failed").toByteArray())
        }
    }

    private fun start(payload: ByteArray, completion: ModuleCompletion) {
        val values = WireMap.decode(payload)
        val kind = ((values["type"] as? WireValue.Integer)?.value ?: 0).toInt()
        val intervalMs = ((values["intervalMs"] as? WireValue.Integer)?.value ?: 100)
            .coerceIn(16, 60_000)
        val platform = when (kind) {
            1 -> Sensor.TYPE_ACCELEROMETER
            2 -> Sensor.TYPE_GYROSCOPE
            3 -> Sensor.TYPE_MAGNETIC_FIELD
            4 -> Sensor.TYPE_ROTATION_VECTOR
            else -> error("Unknown sensor type $kind")
        }
        val sensor = manager.getDefaultSensor(platform) ?: error("Requested sensor is unavailable")
        val id = nextId.getAndIncrement()
        val channel = WatchChannel()
        var lastTimestamp = 0L
        val listener = object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                val timestamp = event.timestamp / 1_000_000L
                if (timestamp - lastTimestamp < intervalMs) return
                lastTimestamp = timestamp
                channel.offer(WireMap.encode(mapOf(
                    "x" to WireValue.Decimal(event.values.getOrElse(0) { 0f }.toDouble()),
                    "y" to WireValue.Decimal(event.values.getOrElse(1) { 0f }.toDouble()),
                    "z" to WireValue.Decimal(event.values.getOrElse(2) { 0f }.toDouble()),
                    "timestamp" to WireValue.Integer(timestamp),
                )))
            }
            override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit
        }
        require(manager.registerListener(listener, sensor, (intervalMs * 1_000).toInt())) {
            "Could not start sensor observation"
        }
        watches[id] = Watch(listener, channel)
        completion.complete(
            ModuleResultStatus.SUCCESS,
            WireMap.encode(mapOf("subscription" to WireValue.Integer(id.toLong()))),
        )
    }

    private fun watch(payload: ByteArray): Watch =
        watches[id(payload)] ?: error("Unknown sensor subscription")

    private fun id(payload: ByteArray): Int =
        ((WireMap.decode(payload)["subscription"] as? WireValue.Integer)?.value
            ?: error("Missing sensor subscription")).toInt()

    private fun stop(id: Int) {
        watches.remove(id)?.let {
            manager.unregisterListener(it.listener)
            it.channel.close()
        }
    }

    override fun close() {
        watches.keys.toList().forEach(::stop)
    }

    private data class Watch(
        val listener: SensorEventListener,
        val channel: WatchChannel,
    )
}

File: ios/Sources/PamNative/Modules/SensorsModule.swift
import CoreMotion
import Foundation

final class SensorsModule: NativeModule, ClosableNativeModule {
    private let lock = NSLock()
    private var nextId = 1
    private var watches: [Int: SensorWatch] = [:]

    func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        do {
            switch method {
            case "watch":
                try start(payload, completion)
            case "next":
                try watch(payload).channel.next(completion)
            case "stop":
                stop(try id(payload))
                completion(.success, Data())
            default:
                throw SensorsModuleError("Unknown sensors method \(method)")
            }
        } catch {
            completion(.failure, Data(error.localizedDescription.utf8))
        }
    }

    private func start(_ payload: Data, _ completion: @escaping ModuleCompletion) throws {
        let values = try WireMap.decode(payload)
        guard case let .integer(kind)? = values["type"] else {
            throw SensorsModuleError("Missing sensor type")
        }
        let intervalMs = min(max(values["intervalMs"]?.integerValue ?? 100, 16), 60_000)
        let manager = CMMotionManager()
        let channel = WatchChannel()
        manager.accelerometerUpdateInterval = Double(intervalMs) / 1_000
        manager.gyroUpdateInterval = Double(intervalMs) / 1_000
        manager.magnetometerUpdateInterval = Double(intervalMs) / 1_000
        manager.deviceMotionUpdateInterval = Double(intervalMs) / 1_000
        let handler: (Double, Double, Double, TimeInterval) -> Void = { x, y, z, timestamp in
            channel.offer((try? WireMap.encode([
                "x": .decimal(x),
                "y": .decimal(y),
                "z": .decimal(z),
                "timestamp": .integer(Int64(timestamp * 1_000)),
            ])) ?? Data())
        }
        switch kind {
        case 1:
            guard manager.isAccelerometerAvailable else { throw SensorsModuleError("Accelerometer is unavailable") }
            manager.startAccelerometerUpdates(to: .main) { value, _ in
                if let value { handler(value.acceleration.x, value.acceleration.y, value.acceleration.z, value.timestamp) }
            }
        case 2:
            guard manager.isGyroAvailable else { throw SensorsModuleError("Gyroscope is unavailable") }
            manager.startGyroUpdates(to: .main) { value, _ in
                if let value { handler(value.rotationRate.x, value.rotationRate.y, value.rotationRate.z, value.timestamp) }
            }
        case 3:
            guard manager.isMagnetometerAvailable else { throw SensorsModuleError("Magnetometer is unavailable") }
            manager.startMagnetometerUpdates(to: .main) { value, _ in
                if let value { handler(value.magneticField.x, value.magneticField.y, value.magneticField.z, value.timestamp) }
            }
        case 4:
            guard manager.isDeviceMotionAvailable else { throw SensorsModuleError("Device motion is unavailable") }
            manager.startDeviceMotionUpdates(to: .main) { value, _ in
                if let value { handler(value.attitude.roll, value.attitude.pitch, value.attitude.yaw, value.timestamp) }
            }
        default:
            throw SensorsModuleError("Unknown sensor type \(kind)")
        }
        lock.lock()
        let id = nextId
        nextId += 1
        watches[id] = SensorWatch(manager: manager, channel: channel)
        lock.unlock()
        completion(.success, try WireMap.encode(["subscription": .integer(Int64(id))]))
    }

    private func watch(_ payload: Data) throws -> SensorWatch {
        let id = try id(payload)
        lock.lock()
        defer { lock.unlock() }
        guard let watch = watches[id] else { throw SensorsModuleError("Unknown sensor subscription") }
        return watch
    }

    private func id(_ payload: Data) throws -> Int {
        let values = try WireMap.decode(payload)
        guard case let .integer(value)? = values["subscription"] else {
            throw SensorsModuleError("Missing sensor subscription")
        }
        return Int(value)
    }

    private func stop(_ id: Int) {
        lock.lock()
        let watch = watches.removeValue(forKey: id)
        lock.unlock()
        watch?.manager.stopAccelerometerUpdates()
        watch?.manager.stopGyroUpdates()
        watch?.manager.stopMagnetometerUpdates()
        watch?.manager.stopDeviceMotionUpdates()
        watch?.channel.close()
    }

    func close() {
        lock.lock()
        let ids = Array(watches.keys)
        lock.unlock()
        ids.forEach { stop($0) }
    }
}

private struct SensorWatch {
    let manager: CMMotionManager
    let channel: WatchChannel
}

private struct SensorsModuleError: LocalizedError {
    let message: String
    init(_ message: String) { self.message = message }
    var errorDescription: String? { message }
}

private extension WireValue {
    var integerValue: Int64? {
        if case let .integer(value) = self { return value }
        return nil
    }
}
